Expand docs with memory & performance info (#1316)
spiceai/docs: website/docs/reference/performance-tuning.md

keywords:
  - performance
  - tuning
  - optimization
  - cayenne
  - duckdb
  - acceleration
pagination_prev: null
pagination_next: null
---

This guide provides recommendations for optimizing Spice performance across data acceleration, query execution, caching, and resource allocation.

## Accelerator Selection

Choose the appropriate [Data Accelerator](/docs/components/data-accelerators) based on dataset characteristics and query patterns.

| Scenario                                | Recommended Accelerator    | Key Configuration                                                     |
| --------------------------------------- | -------------------------- | --------------------------------------------------------------------- |
| Small datasets (<1 GB), lowest latency  | `arrow`                    | Default in-memory                                                     |
| Medium datasets (1-100 GB), complex SQL | `duckdb` with `mode: file` | Set `duckdb_memory_limit`                                             |
| Large datasets (100 GB - 1+ TB)         | `cayenne`                  | Tune cache parameters                                                 |

> | Write-heavy workloads                   | `cayenne` with `zstd`      | Set `cayenne_compression_strategy: zstd`                              |

Reviewer: I don't think this is accurate - `zstd` should only be used when the size on disk is the main concern - setting `zstd` is a tradeoff on query performance to stored size on disk.

### Memory Configuration

```yaml
datasets:
  - from: postgres:schema.table
    name: orders
    acceleration:
      engine: duckdb
      mode: file
      params:
        duckdb_memory_limit: 4GB
        duckdb_file: /data/orders.duckdb
```

**Guidelines:**

- Set `duckdb_memory_limit` to control memory per instance
- DuckDB defaults to 80% of system memory per instance
- Reserve 30% of container memory for the runtime
- Multiple datasets using the same `duckdb_file` share a connection pool

### Connection Pool Tuning

```yaml
acceleration:
  engine: duckdb
  params:
    connection_pool_size: 20  # Default: 10
```

Increase `connection_pool_size` for high-concurrency workloads. Each connection consumes memory.

### Index Configuration

DuckDB supports [ART (Adaptive Radix Tree) indexes](https://duckdb.org/docs/stable/guides/performance/indexing) for faster point lookups:

```yaml
datasets:
  - from: postgres:schema.orders
    name: orders
    acceleration:
      engine: duckdb
      mode: file
      indexes:
        order_id: enabled
        '(customer_id, created_at)': enabled
```

> Indexes consume memory and [do not spill to disk](https://duckdb.org/docs/stable/guides/performance/indexing#indexes-and-memory). Monitor memory usage when adding indexes. For more details on ART index performance, see the [ART paper](https://db.in.tum.de/~leis/papers/ART.pdf).

Reviewer: We should also note that creating the index requires the entire dataset to be loaded into memory.

### Query Parallelism

DataFusion automatically parallelizes queries across available CPU cores. By default, the number of partitions equals the number of CPU cores, providing maximum parallelism.

DataFusion's [FairSpillPool](https://docs.rs/datafusion/latest/datafusion/execution/memory_pool/struct.FairSpillPool.html) divides the configured `memory_limit` evenly among partitions. With more CPU cores, each partition receives less memory, which may increase spilling for memory-intensive queries.

### Join Algorithm Selection

DataFusion supports multiple join algorithms and automatically selects the best one based on query statistics:

| Algorithm        | Memory Usage | Best For                                         |
| ---------------- | ------------ | ------------------------------------------------ |
| Hash Join        | Higher       | Fast execution with sufficient memory (default)  |
| Sort-Merge Join  | Lower        | Memory-constrained environments, pre-sorted data |
| Nested Loop Join | Variable     | Cross joins, non-equi joins                      |

> DataFusion prefers hash joins by default for equi-joins. When memory is constrained and spilling occurs, hash joins may be slower than sort-merge joins due to spill overhead.

Reviewer: hash joins don't spill

### Parquet

Apache Parquet stores data in row groups with per-column statistics. DataFusion uses these statistics to skip row groups that cannot contain matching rows.

**Row Group Pruning:**

Each Parquet row group contains min/max statistics for each column. When a query includes a `WHERE` clause, DataFusion evaluates whether each row group could contain matching rows based on these statistics. Row groups that cannot match are skipped entirely.

For example, with a predicate `WHERE timestamp > '2024-01-01'`, DataFusion skips row groups where the maximum timestamp is before 2024-01-01.

**Page Index:**

Parquet's [Page Index](https://parquet.apache.org/docs/file-format/data-pages/) provides finer-grained statistics at the page level within row groups. DataFusion uses the Page Index to skip individual pages, reducing I/O for selective queries.

**Bloom Filters:**

Parquet files can include bloom filters for membership testing. For equality predicates like `WHERE user_id = 'abc123'`, DataFusion checks the bloom filter before reading column data. If the bloom filter indicates the value is not present, the row group is skipped.

**Late Materialization (Filter Pushdown):**

> DataFusion supports applying filters during Parquet decoding rather than after. This optimization, called late materialization, filters rows before materializing all columns, reducing memory usage for selective queries. Configure via DataFusion settings.

Reviewer: ```suggestion
DataFusion supports applying filters during Parquet decoding rather than after. This optimization, called late materialization, filters rows before materializing all columns, reducing memory usage for selective queries.
```

This isn't relevant for Spice users

### Stale-While-Revalidate

Serve stale cached results while refreshing in the background:

```yaml
runtime:
  caching:
    sql_results:
      enabled: true
      item_ttl: 1m
      stale_while_revalidate_ttl: 5m  # Serve stale for 5m while refreshing
```

This pattern reduces query latency spikes during cache refresh.

## Data Refresh Optimization

### Refresh Mode Selection

| Mode      | Memory Impact | Use Case                                |
| --------- | ------------- | --------------------------------------- |
| `full`    | 2.5x dataset  | Small-medium datasets, complete updates |
| `append`  | Minimal       | Time-series, logs, immutable data       |
| `changes` | Minimal       | CDC-enabled sources                     |
| `caching` | Minimal       | Dynamic content, API data               |

### Append Mode Optimization

Use `time_column` for efficient incremental updates:

```yaml
datasets:
  - from: s3://bucket/events/
    name: events
    time_column: event_time
    acceleration:
      engine: cayenne
      mode: file
      refresh_mode: append
      refresh_check_interval: 5m
```

### Partitioned Data

> For partitioned data, set `time_partition_column` for partition pruning:

Reviewer: ```suggestion
For data that has a granular time column and a separate column for partitioning (i.e. day buckets), set `time_partition_column` to the partitioning column based on time:
```

```yaml
datasets:
  - from: s3://bucket/events/
    name: events
    time_column: event_time
    time_partition_column: event_date  # Physical partition column
    acceleration:
      refresh_mode: append
```

> 

Reviewer: ```suggestion

In this scenario, `event_date` is the day bucket used for physical partitioning (e.g., s3://bucket/events/event_date=2025-10-01/), while event_time provides the granular timestamp for precise filtering.

event_id | event_time (time_column) | event_date (time_partition_column) | event_type | user_id
-- | -- | -- | -- | --
8f2a-1 | 2025-10-01 08:14:22.123 | 2025-10-01 | page_view | u_442
8f2a-2 | 2025-10-01 22:01:05.884 | 2025-10-01 | click | u_901
9c11-a | 2025-10-02 01:12:44.001 | 2025-10-02 | purchase | u_442
9c11-b | 2025-10-02 14:30:12.550 | 2025-10-02 | page_view | u_118
```
